### PR TITLE
feat(cdu): Rework PERF page

### LINKED ISSUE #1823: [BUG] Typo on PERF pages (O/0)
**Mod Version**

master@ 1bbab5800817b94fc74411c234548cb31dc5c90c (latest)

**Describe the bug**

The CLEAN O= speed has a typo on the APPR and GO AROUND phase on the PERF page. It incorrectly uses zero instead of the letter O. It's correct on the TAKE OFF phase.

**To Reproduce**

1. Start a new flight, cold and dark
2. Open the PERF page on the MCDU
3. Go through the various phases and watch the CLEAN O= speed and observe it changing

**Expected behavior**

The letter O should be used on all pages where the CLEAN speed is shown.

**Actual behavior**

TAKE OFF uses the letter O, APPR and GO AROUND use the digit zero.

**References**

PERF TAKE OFF
PERF APPR
PERF GO AROUND

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game? Most likely

Discord username (if different from GitHub): BehEh#1234

### LINKED ISSUE #2096: [BUG] APPR MAG WIND allows invalid values
**Mod Version**

{ "built": "2020-11-24T08:08:38+00:00", "ref": "refs/heads/master", "sha": "b570de8c53cd5b52a21c81f8120bfa57d8c53830", "actor": "wpine215", "event_name": "push" }

**Describe the bug**

The FMC APPR page does not correctly validate and reject invalid headings for the MAG WIND entry.

**To Reproduce**

1. Enter the APPR page
2. Enter random values for the heading of the MAG WIND

**Expected behavior**

Invalid values for heading and speed are rejected

**Actual behavior**

Invalid values for heading are stored, and invalid values for speed are rejected with "NOT ALLOWED"

**References**

image

image

image

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### LINKED ISSUE #3121: [BUG] Certain approaches with long names overlaps with other segments of PERF APPR page on the MCDU
**Mod Version**

Master

{
  "built": "2021-01-20T17:46:20+00:00",
  "ref": "",
  "sha": "",
  "actor": "Imenes <71195324+ImenesFBW@users.noreply.github.com>",
  "event_name": "manual"
}

**Describe the bug**

When selecting certain approaches into the MCDU, such as VORDME approaches, loading into the perf approach page correctly shows the approach in the top right "final" box, however the name is long enough that it overlaps with the F speed.

**To Reproduce**

1. Select an arrival in the flight plan that has a long name, such as a VORDME approach
2. Go to the perf approach page

**Expected behavior**

Name is shortened or smaller so that it doesn't overlap with the F speed.

**Actual behavior**

Name is long enough that is overlaps the F speed.

**References**

Was this working before/when did the issue start occurring?
Unsure

Is this a problem in the vanilla unmodded game?
No, the "final" box text is both smaller and further to the right

Discord username (if different from GitHub): ThatRedMelon#6756

### PR DESCRIPTION
**Pull request by beheh**
Fixes #1823
Fixes #2096
Fixes #3121

## Summary of Changes

This PR started out fixing a minor typo, but turned into a fully fledged makeover for the PERF page.

**Notice: I'm pulling this apart into smaller PRs to make it more manageable.**

It cleans up basic things the default aircraft gets wrong (order of pages, previous pages not accessible) and formats the pages to look much more like the Honeywell original. Note that this PR barely touches the actual functionality of fields, and doesn't touch performance calculations at all. It also does not fix the flight phase transitions (especially around CRZ => CLB => DES) being a bit wonky. This PR is really only about the look and feel of the PERF pages.

Functional changes:
- ~~Fix page order go around (GO AROUND <=> APPR rather than APPR <=> GO AROUND)~~ => #2940
- ~~Disable pages from previous flight phase (e.g. TAKEOFF is no longer accessible during CLB/CRZ/DES etc.)~~ => #2940
- [ ] Fix allowed entries for most fields according to documentation (e.g. -270C° is now rejected)
- [ ] Remember which fields are pilot-entered and which are not (large font/small font)
- [x] Make CI non-editable if no origin/destination is set
- ~~Disable TO SHIFT with dashes if no runway is set~~ => 3082
- ~~Disable various fields during TAKE OFF phase (V speeds, FLAPS, TO SHIFT, THR RED/ACC, ENG OUT ACC, FLX TO)~~ => #3082
- ~~Stop defaulting to flex temp 20C°~~ => #3082
- [ ] Update F/S/O speed calculation (once #2291 lands)

Cosmetic changes:
- [x] Fix typo on APPR and GO AROUND page (letter O instead of zero)
- [x] Fix split text colors for speeds on TAKE OFF, APPR and GO AROUND
- [x] Fix ENG OUT ACC field on GO AROUND
- [x] Fix FLAPS CONF3/FULL text color
- ~~Fix PREV PHASE / NEXT PHASE alignment~~ => #2940
- ~~Fix various label alignments~~ => #2940
- ~~Move take off RWY to header~~ => #3082
- [x] Make QNH amber if within 180nm of destination
- [x] Show amber boxes if CI is not set
- [x] Fix extraneous space in APPR FINAL
- [x] Fix various text colors
- [x] Fix various text sizes
- [x] Color unsupported items inop gray

## Screenshots (if necessary)

image

### Before
grafik

### After
image
image

## References

grafik
grafik
grafik

## Additional context

Discord username (if different from GitHub): BehEh#1234

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by beheh** [comment]
I'm looking if I can easily fix the split white/green text color (O= vs. number) while I'm on this.

**Comment by beheh** [comment]
We had a [short chat in #dev-support](https://discord.com/channels/738864299392630914/747622836381810739/775300500253048852) about how to do split text colors in a nicer way. For now I've just grafted INIT B's method, but I'll try to get the generalized method from https://github.com/flybywiresim/a32nx/blob/e2d8f830bf8bb1a14cebdf35a2dde23ba95e7e47/A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js#L97-L121into a separate PR and can rebase this then afterwards.

**Comment by beheh** [comment]
I'm realizing the flight phases are not correctly implemented, as the aircraft starts in TAKEOFF - until that is fixed (using PREFLIGHT) I cannot disable fields yet.

**Comment by MisterChocker** [comment]
Please adjust your input functions to work with the new input handling #1834

**Comment by MisterChocker** [comment]
Hey I found a "bug". Basically you are unable to insert stuff. And since you are working on the perf page, I thought you may change these to work right.

`A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js`

Lines 671 to 693. 
```js
mcdu.onLeftInput[4] = () => {
            const value = mcdu.inOut;
            mcdu.clearUserInput();
```

```js
mcdu.onRightInput[4] = () => {
            const value = mcdu.inOut;
            mcdu.clearUserInput();
```

remove 

```js
const value = mcdu.inOut;
mcdu.clearUserInput();
```

and add `value` as parameter 
like this:
```js
mcdu.onLeftInput[4] = (value) => {
```

**Comment by beheh** [comment]
@MisterChocker Yep, that was already implicitely fixed as part of this!

**Comment by MisterChocker** [comment]
Oh good to know, thanks!

**Comment by Pushok154** [comment]
Hi, yo have old mcdu style references, here the shot with the new software, with yours version the RWY XX has wrong position and color please take my shot in to account))

fc44bbdd-7d06-42f8-a23a-cb0be4f735df

**Comment by beheh** [comment]
@Pushok154 Thanks, this has already been fixed in master. But photos for all pages would be greatly appreciated!

I also have would need these specific questions answered, or a photo where suitable:

- PERF page before INIT (what is editable? what does the TO SHIFT field look like?) - photo would be great
- What happens if you press LSK 2R with an empty scratchpad (the star)? Does the TO SHIFT change to FT?
- Can FLEX TO TEMP be cleared again after it was entered (CLR => LSK 4R) - or is it NOT ALLOWED?

**Comment by Kimbyeoungjang** [comment]
if you remove transition alt on js, you should remove my code from 32NX\html_ui\Pages\VCockpit\Instruments\Airliners\A320_Neo\PFD\altimeterindicator.js <- there is a trans alt default value(because master was not support default value) search with //Default TRANS ALT to 10,000ft you can find that code

**Comment by Pushok154** [comment]
> * PERF page before INIT (what is editable? what does the TO SHIFT field look like?) - photo would be great

https://drive.google.com/file/d/1-ze5UyAKuBv7fQhG-tPH8KVColEZASBU/view?usp=sharing

> * What happens if you press LSK 2R with an empty scratchpad (the star)? 
> * Can FLEX TO TEMP be cleared again after it was entered (CLR => LSK 4R) - or is it NOT ALLOWED?

https://drive.google.com/file/d/1v__byaJj4SQQ7YLLsU7A6st-OVMTSjQV/view?usp=sharing

> * Does the TO SHIFT change to FT?
This is feature firmware and pilots have no opportunities to change ft meters)

> * (the star)?

Снимок экрана 2020-11-26 в 17 35 18

> * But photos for all pages would be greatly appreciated!

IMG_4894

IMG_4895

IMG_4896

IMG_4897

**Comment by wpine215** [comment]
@beheh Any update on the last checkbox item in this PR? Looks almost ready to merge to me.

**Comment by beheh** [comment]
Not quite yet, sorry :< I need to systematically go through to ensure all is. Also still missing the large/small font. I want to see this in QA asap as well, it's been hanging over me. I'll really try to get it done this week.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js [hunk 1]
```diff
@@ -1023,20 +1024,22 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
...
-        }
-        if (this.perfTOTemp <= 45) {
-            return 92.2;
-        }
-        if (this.perfTOTemp <= 50) {
-            return 90.5;
-        }
-        if (this.perfTOTemp <= 55) {
-            return 88.8;
+        if (isFinite(this.perfApprTransAlt)) {
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:1027]: shouldn't this check for `this.perfTOTemp`?

**Comment by MisterChocker** [comment]
Basically C&D with no data added.
This might be some very nice details.
image
image

**Comment by beheh** [comment]
That's Thales, based on the font (serifs on the I). But thanks! These can help to figure out some of the details.

**Comment by beheh** [comment]
Heads up to people watching this: As this PR has certainly ballooned and gotten bit large and daunting, I'm pulling it apart into smaller, more manageable PRs. Then things can start landing, and people can see more clearly what has actually changed and don't feel blocked to work on unrelated parts of the PERF pages.

**Comment by beheh** [comment]
The page order fixes and reworked TAKE OFF page have been landed, the remaining page work (especially APPR) is being taken over by @tracernz. Closing this for now.

@tracernz: Make sure to pick up the fixed issues above when opening your PR containing the APPR page.

**Pull request closed without merging**